File: driftwatch/backend/internal/alerts/discord_test.go
package alerts

import (
	"encoding/json"
	"io"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"

	"github.com/YOURUSERNAME/driftwatch/internal/agent"
	"github.com/YOURUSERNAME/driftwatch/internal/gemini"
)

func TestSendDriftAlert_PostsExpectedEmbed(t *testing.T) {
	var captured discordPayload

	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		require.Equal(t, http.MethodPost, r.Method)
		require.Equal(t, "application/json", r.Header.Get("Content-Type"))
		body, err := io.ReadAll(r.Body)
		require.NoError(t, err)
		require.NoError(t, json.Unmarshal(body, &captured))
		w.WriteHeader(http.StatusNoContent)
	}))
	defer srv.Close()

	c := NewClient(srv.URL)

	result := &gemini.AnalysisResult{
		Severity:    agent.SeverityCritical,
		Summary:     "Web container is gone in prod",
		FixCommand:  "docker compose up -d web",
		Explanation: "The web service is declared but not running.",
	}
	drifts := []agent.DriftEvent{
		{Type: agent.DriftTypeMissingContainer, ContainerName: "web", Severity: agent.SeverityCritical},
		{Type: agent.DriftTypeEnvMismatch, ContainerName: "web", Severity: agent.SeverityCritical},
		{Type: agent.DriftTypeImageStale, ContainerName: "api", Severity: agent.SeverityWarning},
	}

	err := c.SendDriftAlert("acme-prod", result, drifts)
	require.NoError(t, err)

	require.Len(t, captured.Embeds, 1)
	e := captured.Embeds[0]
	assert.Equal(t, "⚠️ Drift Detected — acme-prod", e.Title)
	assert.Equal(t, 15158332, e.Color)
	require.Len(t, e.Fields, 5)

	byName := make(map[string]discordField, len(e.Fields))
	for _, f := range e.Fields {
		byName[f.Name] = f
	}

	sev := byName["Severity"]
	assert.Equal(t, "critical", sev.Value)
	assert.True(t, sev.Inline, "Severity should be inline")

	count := byName["Containers Affected"]
	assert.Equal(t, "2", count.Value, "web+api are the two unique containers")
	assert.True(t, count.Inline, "Containers Affected should be inline")

	assert.Equal(t, "Web container is gone in prod", byName["Summary"].Value)
	assert.False(t, byName["Summary"].Inline)

	fixField := byName["Fix Command"].Value
	assert.Contains(t, fixField, "```bash")
	assert.Contains(t, fixField, "docker compose up -d web")
	assert.Contains(t, fixField, "```")

	assert.Equal(t, "The web service is declared but not running.", byName["Explanation"].Value)

	assert.Contains(t, e.Footer.Text, "DriftWatch")
	assert.Contains(t, e.Footer.Text, "detected at")
	assert.NotEmpty(t, e.Timestamp)

	// Timestamp should parse as RFC3339.
	_, err = time.Parse(time.RFC3339, e.Timestamp)
	assert.NoError(t, err)
}

func TestSendDriftAlert_ColorBySeverity(t *testing.T) {
	cases := []struct {
		sev      string
		expected int
	}{
		{agent.SeverityCritical, 15158332},
		{agent.SeverityWarning, 16776960},
		{agent.SeverityInfo, 3447003},
	}
	for _, c := range cases {
		assert.Equal(t, c.expected, colorForSeverity(c.sev))
	}
}

func TestSendDriftAlert_NonOKStatusReturnsError(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusBadRequest)
		_, _ = w.Write([]byte(`{"message":"invalid embed"}`))
	}))
	defer srv.Close()

	c := NewClient(srv.URL)
	result := &gemini.AnalysisResult{Severity: agent.SeverityInfo, FixCommand: "noop"}
	err := c.SendDriftAlert("p", result, nil)
	require.Error(t, err)
	assert.Contains(t, err.Error(), "400")
	assert.Contains(t, err.Error(), "invalid embed")
}

func TestSendDriftAlert_NilResultIsError(t *testing.T) {
	c := NewClient("https://example.invalid")
	err := c.SendDriftAlert("p", nil, nil)
	require.Error(t, err)
	assert.Contains(t, err.Error(), "nil")
}

func TestShouldAlert_Cooldown(t *testing.T) {
	c := NewClient("https://example.invalid")

	assert.True(t, c.ShouldAlert("p", nil), "nil lastAlertedAt should allow alert")

	recent := time.Now().Add(-5 * time.Minute)
	assert.False(t, c.ShouldAlert("p", &recent), "5-min-old alert should be in cooldown")

	just := time.Now().Add(-29 * time.Minute)
	assert.False(t, c.ShouldAlert("p", &just), "29-min-old alert should still be in cooldown")

	old := time.Now().Add(-31 * time.Minute)
	assert.True(t, c.ShouldAlert("p", &old), "31-min-old alert should allow new alert")
}

File: driftwatch/backend/src/services/discord.test.ts
import { describe, expect, it, vi } from 'vitest';

import { DriftType, Severity, type DriftEvent } from '../types.js';
import { colorForSeverity, DiscordClient, type DiscordPayload } from './discord.js';
import type { AnalysisResult } from './gemini.js';

function drift(
	type: string,
	containerName: string,
	severity: string
): DriftEvent {
	return {
		id: 'evt',
		type: type as DriftEvent['type'],
		containerName,
		liveValue: '',
		declaredValue: '',
		severity: severity as DriftEvent['severity'],
		detectedAt: new Date()
	};
}

/** Records the outgoing request and replies with the given status. */
function captureFetch(status = 204, body = '') {
	const calls: { url: string; init: RequestInit }[] = [];
	const impl = vi.fn(async (url: unknown, init: unknown) => {
		calls.push({ url: String(url), init: init as RequestInit });
		// A 204 is a null-body status; an empty string still counts as a body.
		return new Response(body || null, { status });
	});
	return { calls, impl: impl as unknown as typeof fetch };
}

describe('DiscordClient', () => {
	it('posts the expected embed', async () => {
		const { calls, impl } = captureFetch();
		const c = new DiscordClient('https://discord.test/webhook', { fetchImpl: impl });

		const result: AnalysisResult = {
			severity: Severity.Critical,
			summary: 'Web container is gone in prod',
			fixCommand: 'docker compose up -d web',
			explanation: 'The web service is declared but not running.',
			driftBreakdown: []
		};
		const drifts = [
			drift(DriftType.MissingContainer, 'web', Severity.Critical),
			drift(DriftType.EnvMismatch, 'web', Severity.Critical),
			drift(DriftType.ImageStale, 'api', Severity.Warning)
		];

		await c.sendDriftAlert('acme-prod', result, drifts);

		expect(calls).toHaveLength(1);
		expect(calls[0]!.init.method).toBe('POST');
		expect(
			(calls[0]!.init.headers as Record<string, string>)['Content-Type']
		).toBe('application/json');

		const payload = JSON.parse(calls[0]!.init.body as string) as DiscordPayload;
		expect(payload.embeds).toHaveLength(1);

		const e = payload.embeds[0]!;
		expect(e.title).toBe('⚠️ Drift Detected — acme-prod');
		expect(e.color).toBe(15158332);
		expect(e.fields).toHaveLength(5);

		const byName = new Map(e.fields.map((f) => [f.name, f]));
		expect(byName.get('Severity')).toMatchObject({ value: 'critical', inline: true });
		// web + api are the two unique containers.
		expect(byName.get('Containers Affected')).toMatchObject({
			value: '2',
			inline: true
		});
		expect(byName.get('Summary')!.value).toBe('Web container is gone in prod');
		expect(byName.get('Summary')!.inline).toBeUndefined();
		expect(byName.get('Fix Command')!.value).toContain('```bash');
		expect(byName.get('Fix Command')!.value).toContain('docker compose up -d web');
		expect(byName.get('Explanation')!.value).toBe(
			'The web service is declared but not running.'
		);

		expect(e.footer.text).toContain('DriftWatch');
		expect(e.footer.text).toContain('detected at');
		expect(Number.isNaN(Date.parse(e.timestamp))).toBe(false);
	});

	it('maps severity to colour', () => {
		expect(colorForSeverity(Severity.Critical)).toBe(15158332);
		expect(colorForSeverity(Severity.Warning)).toBe(16776960);
		expect(colorForSeverity(Severity.Info)).toBe(3447003);
		expect(colorForSeverity('nonsense')).toBe(3447003);
	});

	it('throws on a non-OK status, including the response body', async () => {
		const { impl } = captureFetch(400, '{"message":"invalid embed"}');
		const c = new DiscordClient('https://discord.test/webhook', { fetchImpl: impl });

		await expect(
			c.sendDriftAlert('p', {
				severity: Severity.Info,
				summary: '',
				fixCommand: 'noop',
				explanation: '',
				driftBreakdown: []
			}, [])
		).rejects.toThrow(/400.*invalid embed/s);
	});

	it('throws when the analysis is null', async () => {
		const c = new DiscordClient('https://discord.test/webhook');
		await expect(c.sendDriftAlert('p', null, [])).rejects.toThrow(/nil/);
	});

	it('is a no-op when no webhook is configured', async () => {
		const { calls, impl } = captureFetch();
		const c = new DiscordClient('', { fetchImpl: impl });

		await expect(c.sendDriftAlert('p', null, [])).resolves.toBeUndefined();
		expect(calls).toHaveLength(0);
	});

	it('enforces a 30 minute alert cooldown', () => {
		const c = new DiscordClient('https://discord.test/webhook');
		const minutesAgo = (n: number) => new Date(Date.now() - n * 60_000);

		expect(c.shouldAlert('p', null)).toBe(true);
		expect(c.shouldAlert('p', minutesAgo(5))).toBe(false);
		expect(c.shouldAlert('p', minutesAgo(29))).toBe(false);
		expect(c.shouldAlert('p', minutesAgo(31))).toBe(true);
	});
});
